MultiQubit_TimeDomain_Analysis: cleanup and improvements

stephlazar opened the pull request:
With this commit, this class will work when CalibrationPoints are not given or are empty, and when data_to_fit is None or {}
Also contains code reorganization + wrapping chunks into methods.

Commit:
MultiQubit_TimeDomain_Analysis: cleanup and improvements

- this class now works when CalibrationPoints are not given or are empty, and when data_to_fit is None or {}
- code reorganization + wrapping chunks into methods

Commit:
MultiQubit_TimeDomain_Analysis: if no cal points are found, default rotation_type to "global_PCA" if TwoD + fix bug for no cal points info given and 3 cal points measured

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -289,11 +289,66 @@ def extract_data(self):
 
-        self.data_to_fit = deepcopy(self.get_param_value('data_to_fit', {}))
+        # whether to rotate data or not
+        self.rotate = self.get_param_value('rotate', default_value=True)
```
Was it intended that you changed the default to True? (not indicated in commit message or PR)

stephlazar replied:
Yes, I thought doing this makes sense. I will still need to test if anythings breaks

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -697,10 +708,28 @@ def unique(l):
+                        self.num_cal_points,
+                        self.proc_data_dict['sweep_points_dict'][qbn][
+                            'sweep_points'])} for qbn in self.qb_names}
+            self.proc_data_dict['sweep_points_dict'] = sweep_points_w_calpts
```
Modifying self.proc_data_dict['sweep_points_dict'] does not fit to the name of the method.

It might also be helpful to explain in a docstring what the method does, or at least in a comment at the point where the method is called.

stephlazar replied:
this was now cleaned up

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -349,6 +436,13 @@ def create_sweep_points_dict(self):
+                {qbn: {'sweep_points': self.cp.extend_sweep_points(
+                    self.proc_data_dict['sweep_points_dict'][qbn][
+                        'sweep_points'], qbn)} for qbn in self.qb_names}
+            self.proc_data_dict['sweep_points_dict'] = sweep_points_w_calpts
```
What is the difference (or the different use case) between these lines of code and the respective code in get_num_cal_points?

Defining the local sweep_points_w_calpts might be unneeded.

stephlazar replied:
Now this is done in the new method update_sweep_points_dict(). Before, get_cal_points was extending sweep_points by the number of cal points from self.cp (if it exists), but then get_num_cal_points() did it again if it found that cal point info was not provided by the measurement did contain cal points.

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -289,11 +289,66 @@ def extract_data(self):
+            if isinstance(self.data_to_fit[qbn], (list, tuple)):
+                self.data_to_fit[qbn] = self.data_to_fit[qbn][0]
+
+        # get data_filter
```
Should this go into a method?

stephlazar replied:
Maybe? :D
I will try to clean it up more here and the self.data_to_fit chunk above it

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -289,11 +289,66 @@ def extract_data(self):
+        # creates self.cp if found else defaults to legacy extraction
+        self.get_cal_points()
+
+        # get data_to_fit
```
Should this go into a method?

stephlazar replied:
See above reply

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -309,6 +364,38 @@ def get_sweep_points(self):
+                last_ge_pulses, self.qb_names[0])[self.qb_names[0]] if \
+                self.rotate else cal_states_rotations
+            self.cal_states_rotations = self.get_param_value(
+                'cal_states_rotations', default_value=cal_states_rots)
```
It is a bit confusing to have cal_states_rotations, cal_states_rots, and self.cal_states_rotations. Is it possible to use names that make the purpose of each variable clearer?

stephlazar replied:
Yes probably makes sense, but I don't really want to clean this up now (need to carefully go through the code to change everywhere where we set them in the metadata as well). But if you have a strong opinion that I should do it now, let me know.

chellings replied:
Then keep the variable names, but maybe add comments here to make clear what happens and what the names mean?

stephlazar replied:
👍

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -309,6 +364,38 @@ def get_sweep_points(self):
+                if self.rotate else cal_states_rotations
+            self.cal_states_dict = self.get_param_value('cal_states_dict',
+                                                        default_value={})
+        if self.cal_states_rotations is None:
```
```
if self.cal_states_rotations is None:  # if an error happens above
```

correct?

stephlazar replied:
Not really, it can be that somehow it was None in the metadata. This line essentially gives this attribute an expected default value. We could raise a warning here to let the user know None is not accepted.

chellings replied:
if self.cal_states_rotations is None:
    # if an error happens above or None was provided in the metadata

I can't judge whether it is a common thing that the user writes None into the metadata and whether we need to warn them about it. I just wanted a comment to make clear for a reader of the code why we end up in this branch.

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -289,11 +289,66 @@ def extract_data(self):
+        if self.data_to_fit is None:
+            # If we have cal points, but data_to_fit is not specified,
+            # choose a reasonable default value. In cases with only two cal
+            # points, this decides which projected plot is generated. (In
```
I thought the plan was that data_to_fit should no longer decide about which projected plots are created?

stephlazar replied:
Not yet. Nathan implemented this in the preselection_for_tda PR I think. Or his active reset branch.

chellings replied:
if it was on the tda branch, it should already be part of S17. Concerning the active reset branch: please coordinate with nathan how to get rid of this branch. Since it has de facto become the new S17 measurement branch, it does not make sense to keep his active reset branch separate. And it also does not make sense to clean up tda in parallel on two branches. Since your PR is a PR into S17 anyways, you should probably directly combine it with Nathan's cleanup, and merge both (active reset and this PR) into S17.

stephlazar replied:
I told Nathan on Wed that he should try to merge his branch into S17. And that he should handle the preselection_for_ta PR. And yes makes sense to coordinate this cleanup effort.

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -309,6 +364,38 @@ def get_sweep_points(self):
+                                                        default_value={})
+        if self.cal_states_rotations is None:
+            self.cal_states_rotations = cal_states_rotations
+        if not len(self.cal_states_dict):
```
Please add a comment to explain what happens here.

Is the case of rotate=False and/or single shot treated properly here?

Is the case of column/row PCA treated properly here?

I am generally a bit worried that the fact that a method for getting cal points changes the rotation type might have unexpected side effects.

stephlazar replied:
Good catch! This is indeed a bug. I fixed it now by wrapping rotation_type extraction into a method.

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -429,103 +523,24 @@ def create_meas_results_per_qb(self):
         self.create_sweep_points_dict()
         self.create_meas_results_per_qb()
-
-        # temporary fix for appending calibration points to x values but
```
Just wondering: you removed this comment when moving the part of code. Does this mean that the temporary fix has now become the long-term solution?

stephlazar replied:
I removed it because it seemed like a merge bug. The comment wasn't relevant to the lines of code above or below. It seems like it used to belong to the cal points extraction lines. But now I don't find any temporary fix of this sort and in this cleanup effort I make sure we always handle correctly this extension based on cal points (or at least I hope so ... )

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -289,11 +289,66 @@ def extract_data(self):
+        self.get_cal_points()
+
+        # get data_to_fit
+        self.data_to_fit = deepcopy(self.get_param_value('data_to_fit'))
```
Is such a deepcopy also needed for other parameters, e.g. for cal points (see below)?

stephlazar replied:
It is indeed needed for the parameters that we then modify. So I now deepcopy data_to_fit and rotation_type.

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -289,11 +289,66 @@ def extract_data(self):
+                self.data_to_fit = {qbn: 'pg' for qbn in self.qb_names}
+            else:
+                self.data_to_fit = {}
+        # TODO: Steph 15.09.2020
```
Are there plans to resolve this TODO?

stephlazar replied:
No :D This is only used by the cphase analysis (I needed to hack this when Christian wanted to get the amount of swap errors and he needed to fit multiple projected data for the same qubit). It would break all the analyses we have and it will be a lot of work to upgrade them. Maybe at some point when we have a further use case.

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -429,103 +523,24 @@ def create_meas_results_per_qb(self):
             # this assumes data obtained with classifier detector!
             # ie pg, pe, pf are expected to be in the value_names
+            if not len(self.data_to_fit):
+                self.data_to_fit = {qbn: 'pe' for qbn in self.qb_names}
```
Why is this a reasonable data_to_fit not yet ensured at this point? E.g. in the part "# get data_to_fit" in line 318?

If it is really needed to do this late point, please add a comment that explains why.

stephlazar replied:
Yes, can be removed now and I have done so.

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -697,10 +708,28 @@ def unique(l):
+        mrpq = self.proc_data_dict['meas_results_per_qb']
+        mrpq_raw_dict = mrpq[list(mrpq)[0]]
+        num_data_points = len(mrpq_raw_dict[list(mrpq_raw_dict)[0]])
+        if self.num_cal_points == 0 and num_data_points != num_sp:
```
When is this condition expected to be True? Please add a comment

stephlazar replied:
Comment added. It is True when no calibration points information was given (cal_points in options dict or metadata is None or an empty CalibrationPoints instance), but the measurement did contain calibration segments. We want to exclude the data corresponding to them from the fit.

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -719,9 +748,16 @@ def get_cal_data_points(self):
-                self.cal_states_dict_for_rotation[qbn][cal_state] = \
-                    None if do_PCA and self.num_cal_points != 3 else \
-                        self.cal_states_dict[cal_state]
+                if do_PCA and self.num_cal_points != 3:
```
What is special in the case of 3 cal points? Please add a comment.

stephlazar replied:
Before reading this, I had added the comment "Cannot do PCA for 3 cal states." Is this just not implemented or cannot be done conceptually?

chellings replied:
Not sure I understand this. Why does the number of cal states matter for PCA? I thought the point of PCA is to ignore the cal states and rotate based on all recorded data.

stephlazar replied:
This was fixed in the meantime. Now it just checks whether do_PCA is True and the cal_states_dict for that qubit is empty

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -719,9 +748,16 @@ def get_cal_data_points(self):
+                    if not len(self.cal_states_dict):
+                        self.cal_states_dict_for_rotation[qbn][cal_state] = None
+                        if self.get_param_value('TwoD', default_value=False):
+                            self.rotation_type = 'global_PCA'
```
Why is it needed that a method with name get_cal_data_points changes the rotation type? Can this be avoided by restructuring?

stephlazar replied:
Yes I did that now.

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -719,9 +748,16 @@ def get_cal_data_points(self):
+                        self.cal_states_dict_for_rotation[qbn][cal_state] = \
+                            self.cal_states_dict[cal_state]
 
     def cal_states_analysis(self):
```
Method name not self-explanatory. A docstring might help.

stephlazar replied:
I now changed the name and added a docstring.

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -778,7 +816,7 @@ def cal_states_analysis(self):
                             column_PCA=self.rotation_type == 'column_PCA'))
             else:
-                if len(cal_states_dict) == 3:
+                if len(cal_states_dict) == 3 and 'PCA' not in self.rotation_type:
```
Shouldn't self.num_cal_points be used here?

What happens in case of 3 cal points and PCA?

In other instances you convert to lower case when checking for PCA

stephlazar replied:
No, we shouldn't use self.num_cal_points because this attribute counts how many calibration segments were used in the experiment. So it will be nonzero even if no calibration point information was provided (reflected by self.cal_states_dict = {qbn: {} for qbn in self.qb_names}. Here we want to know if it should do 3 state rotation, which it should only do if this instruction was given (contained in self.cal_states_dict and self.cal_states_rotations). 

I'm not sure where 3 cal points and PCA is handled :D I have to still look into this.

Yes, I now tried to standardize this check such that we always convert to .lower().

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -6012,6 +6049,8 @@ def process_data(self):
                 sweep_points, self.proc_data_dict['sweep_points_dict'][qbn][
                     'cal_points_sweep_points']])
 
+            self.get_num_cal_points()
```
Hasn't this already been called in process_data? It might help to add a comment why this needs to be called again.

stephlazar replied:
I now cleaned this up and fixed bugs.

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -1640,7 +1676,8 @@ def prepare_projected_data_plot(
         plot_names_cal = []
-        if plot_cal_points and self.num_cal_points != 0:
+        if plot_cal_points and self.num_cal_points != 0 and \
+                len(self.cal_states_dict):
```
What is the difference between num_cal_points and the len of the cal states dict?

stephlazar replied:
I answered this in the context of another question but here it is again.

num_cal_points counts how many calibration segments were used in the experiment. So it will be nonzero even if no calibration point information was provided (reflected by self.cal_states_dict = {qbn: {} for qbn in self.qb_names}).

chellings replied:
OK, thanks. Maybe one could add a comment here to state the intuitive meaning of the condition for this branch of the if.

stephlazar replied:
Added now

chellings reviewed (changes requested):
Thanks a lot for starting this cleanup.

Since there are many things that I do not fully understand about this code, I have just written down everything that came to my mind. I hope that this will help to continue the cleanup with the aim of converging to code that even I understand ;-).

This seems to touch the core of the tda with the potential of unexpected side effects on all sorts of measurements. Please provide a list for which kind of measurements you have run tests (including what you passed for the relevant parameters, e.g., rotation type, cal points, and data_to_fit). I would send you suggestions of further measurements and parameter combinations that should be tested.

You introduce various methods that are called "get_..." which might be a bit confusing since they do not return anything, but only preprocess properties of self. Would some other name be more appropriate? E.g. "process_..."?

Commit:
Merge branch 'devel/S17/timedomain_classes' into enh/tda_cleanup

Commit:
Merge branch 'devel/S17/timedomain_classes' into enh/tda_cleanup

Commit:
Merge branch 'devel/S17/timedomain_classes' into enh/tda_cleanup

Commit:
Merge branch 'devel/S17/timedomain_classes' into enh/tda_cleanup

Commit:
Merge branch 'devel/S17/timedomain_classes' into enh/tda_cleanup

Commit:
Merge branch 'devel/S17/timedomain_classes' into enh/tda_cleanup

Commit:
timedomain_analysis.py: more cleanup throughout the file, with a focus in MultiQubit_TimeDomain_Analysis and MultiCZgate_Calib_Analysis

Highlights
- cal_states_dict and cal_states_rotations are now always dicts of dicts with qubit names the outermost keys; necessary changes done throughout the file
- cleanup of how self.data_to_fit is handled: logic from cal_states_analysis moved to extract_data; self.data_to_fit is then updated in cal_states_analysis()
- fixes bugs in MultiCZgate_Calib_Analysis.process_data introduced by changes in the base class. MultiCZgate_Calib_Analysis and its children can now handle PCA
- fixes bugs in get_transition_name

Commit:
QScaleAnalysis: (bugfix) removes call to self.get_num_cal_points which made analysis crash when no cal points are given

Commit:
timedomain_analysis.py: major enhancement + bugfix in QScaleAnalysis

Highlights:
 - fixes remaining bug in QScaleAnalysis that broke backwards compatibility with the qscale measurement implemented in multi_qubit_module.py and QuDev_transmon.measure_qscale() (sweep points need to be repeated 3x in the analysis)
  - introduces sp_1d_filter (used in MultiQubit_TimeDomain_Analysis.create_sweep_points_dict()
  - moves calls to self.create_sweep_points_dict(), self.create_meas_results_per_qb(), self.get_num_cal_points(), self.update_sweep_points_dict() from MultiQubit_TimeDomain_Analysis.process_data() to self.extract_data()
 - adds MultiQubit_TimeDomain_Analysis.update_sweep_points_dict() to clean up sweep_points_dict creation (removes lines dealing with this from self.get_cal_points() and self.process_data())
 - renames the MultiQubit_TimeDomain_Analysis methods
  - self.get_cal_data_points --> self.get_cal_states_dict_for_rotation
  - self.cal_states_analysis --> self.rotate_and_project_data
 - adds docstrings for most methods of MultiQubit_TimeDomain_Analysis

stephlazar commented:
I did another major restructuring of the code in the meantime and I still need to commit more fixed after going through your comments :p 
I have been testing with Rabi, QScale, Cphase, DynPhase, Chevron. It would make sense to compile a list of what should be tested and that you give me the relevant calls + timestamps for the things that I'm not so familiar with (RabiFreqSweep, flux pulse scope (also with split data), drive xtalk cancellation, T1 freq sweep (ie defect mode spec)).

Commit:
timedomain_analysis.py: more cleanup

Highlights:
 - adds get_data_to_fit, get_rotation_type, get_data_filter to MultiQubit_TimeDomain_Analysis
 - cleans up get_cal_states_dict_for_rotation and rotate_and_project_data
 - changes rotation_type to be a dict per qubit (string is still accepted as input)
 - figure names and titles now contain type of PCA that was done
 - docstring improvements

Commit:
Merge branch 'devel/S17/timedomain_classes' into enh/tda_cleanup

Commit:
timedomain_analysis.py: wraps channel_map generation code in self.get_channel_map + moves data_filter extraction line from extract_data to get_data_filter

Commit:
Merge branch 'devel/S17/timedomain_classes' into enh/tda_cleanup

Commit:
Merge branch 'devel/S17/timedomain_classes' into enh/tda_cleanup

Commit:
Merge branch 'devel/S17/timedomain_classes' into enh/tda_cleanup

stephlazar commented:
I had no problems merging the active reset branch. In the end I decided to keep the changes that Nathan had implemented there. So now you can re-review this PR (does not include active_reset).

Commit:
Merge branch 'devel/S17/timedomain_classes' into enh/tda_cleanup

Commit:
More cleanup and minor improvements to plotting:

- data_to_fit[qbn] as suffix in figure names
 - added in Rabi, RabiFrequencySweep, Ramsey, ReparkingRamsey, QScale, T1, T1FrequencySweep
- only plot pe = 0.5 line in Ramsey if pca was not done
- T1FrequencySweepAnalysis: add cbar label

stephlazar commented:
@chellings 
Branch was now successfully tested with the following analyses

- Rabi (ge + ef)
- RabiFrequencySweep
- Ramsey
- RamseyReparking
- QScale (mqm + qe versions)
- T1
- Echo
- Flux pulse scope
- CPhase + DynPhase
- T1FrequencySweep
- Chevron

The tests included

- data_to_fit = None or {}
- cal_points = None or empty CalibrationPoints object
- various rotation types

Commit:
EchoAnalysis: bugfix (define self.qb_names) + add data_to_fit suffix to figure name

Commit:
Merge branch 'devel/S17/timedomain_classes' into enh/tda_cleanup

Commit:
Merge branch 'devel/S17/timedomain_classes' into enh/tda_cleanup

Commit:
Merge branch 'feature/timedomain_classes' into enh/tda_cleanup

# Conflicts:
#	pycqed/analysis_v2/timedomain_analysis.py

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -255,6 +262,35 @@ def __init__(self,
+            - self.rotate: bool; whether to do data rotation and projection.
+                Must be False if raw data is already classified.
+            - self.predict_proba: bool (only used for SSRO data);
+                whether to classify shots
```
Some of those seem to be created in called methods (e.g., data_filter, channel_map). Others I could not find at all in extract_data (measurement_strings).

stephlazar replied:
I removed the ones that are created in methods. measurement_strings is a few lines below

chellings replied:
Sorry about the measurement_strings. It was hidden because it was unchanged

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -266,15 +302,89 @@ def extract_data(self):
+        # to be used if data_type == singleshot (see process_data)
+        self.predict_proba = self.get_param_value("predict_proba", False)
+        if self.predict_proba and self.get_param_value("classified_ro", False):
+            log.warning("predict_proba set to 'False' as probabilities are"
```
space missing at the line break in the string

stephlazar replied:
👍

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -266,15 +302,89 @@ def extract_data(self):
+        if self.options_dict.get('TwoD', False):
+            self.create_sweep_points_2D_dict()
+
+    def get_params_from_file(self, params_dict=None, numeric_params=None):
```
Is this still needed? It seems that it is no longer used in any of the child classes.

stephlazar replied:
No, it's not. Removed it now.

stephlazar replied:
It isn't. I removed it now

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -290,17 +400,267 @@ def extract_data(self):
+                be identical for each qubit, since each calibration point is a
+                separate segment and the measurement cannot be configured to
+                have different number of segments for different qubit in the
+                same measurement.
```
But what if you have 2-qb cal points gg, ge, eg, ee? Then, the number of segments is the same for each qubit, but the inner dicts would not be identical. (see CalibrationPoints.multi_qubit with all_combinations=True)

stephlazar replied:
There is currently no support for this in this module as far as I know. So maybe we can allow this in a later feature? Maybe I can just update the comment here.

stephlazar replied:
I think what I wanted to say here was that the inner dicts must have the same number of entries but the values (ie the lists specifying the indices) can be different. 

I don't know, I need to try haha

stephlazar replied:
I'm not sure if this class will work with that case. One would need to test it. 
I updated the comment after trying a few things. It seems the inner dicts must contain all the cal points in the CalibrationPoints object. Essentially cal_states_dict is CalibrationPoints.get_indices() ...
cal_states_rotations then decides what kind of rotation is done for which qubit.

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -290,17 +400,267 @@ def extract_data(self):
+        cal_states_rotations and cal_states_dict.
+
+        ! Updates self.rotation_type if no calibration points information w
+        was found !
```
Outdated?

stephlazar replied:
Yes

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -290,17 +400,267 @@ def extract_data(self):
+        try:
+            # try to take cal points information from CalibrationPoints instance
+            # or its repr
+            self.cp = CalibrationPoints.from_string(cal_points)
```
In the future (not necessarily in this PR), we might want to get rid of the from_string and do it like for SweepPoints

stephlazar replied:
Yes, we should do that at some point.

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -290,17 +400,267 @@ def extract_data(self):
+            log.warning('Parameter "global_PCA" is deprecated. Please set '
+                        'rotation_type="global_PCA" instead.')
+        # Get the rotation_type. See class docstring
+        # Without deepcopy, the value in the options_dict/metadata will be
```
would be

stephlazar replied:
👍

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -290,17 +400,267 @@ def extract_data(self):
+
+        """
+        # Without deepcopy, the value in the options_dict/metadata will be
+        # modified as well. This will break MultiCZgate_Calib_Analysis which
```
will --> would (2 times)

stephlazar replied:
👍

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -290,17 +400,267 @@ def extract_data(self):
+        if self.data_to_fit is None or not len(self.data_to_fit):
+            # If we have cal points, but data_to_fit is not specified,
+            # choose a reasonable default value. In cases with only two cal
+            # points, this decides which projected plot is generated. (In
```
I am not sure whether I remember it correctly, but I thought we had at some point discussed to separate the decision which plots are generated from the question which data is the basis for fitting. Wasn't this part of the motivations of the whole cleanup? Was this somehow related to the new storing_keys? Maybe I am mixing up something. We could also discuss orally.

BTW enh/tda_storing_keys could become an MR once enh/preselection_for_tda got merged

stephlazar replied:
I don't remember if this was part of the motivation for the cleanup (I think it was rather triggered by the fact the analysis was failing if data_to_fit was {} or None or if cal points were not specified). In any case, this comment is outdated. The plots are generated based on what is inside self.proc_data_dict['projected_data_dict']. Before Nathan added the "other_cs" feature to store both pe and pg=1-pe, only pe used to be stored (typically decided by data_to_fit)

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -290,17 +400,267 @@ def extract_data(self):
+                    self.data_to_fit[qbn] = f'p{csr[-1][0]}'
+
+        # TODO: Steph 15.09.2020
+        # This is a hack to allow list inside data_to_fit.
```
open TODO item (and what is this needed for?)

stephlazar replied:
This is only needed by the MultiCZgate_Calib_Analysis to allow the same data to be fitted in two ways (to extract SWAP errors). We just needed it for this specific case and I didn't want to go change the entire module just for this (it would break all children who assume the values are strings not lists).

chellings replied:
Yeah, you mentioned something like that, but this explanation should probably go into the comment in the code to avoid that this question comes up again :-)

stephlazar replied:
👍

stephlazar replied:
Yes, the explanation is there now

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -688,48 +987,146 @@ def unique(l):
+
+        Creates self.proc_data_dict['projected_data_dict'].
+
+        ! Note: self.data_to_fit is updated to reflect what was doe here for
```
typo: doe

I have not fully understood why the decision about the type of rotation is not already fixed when we are in extract_data. Also I wonder whether there is a risk that someone will in the future rely on data_to_fit begin usable after the call to get_data_to_fit and then introduce bugs because data_to_fit still gets modified afterwards. Do these thoughts make sense?

stephlazar replied:
It was a convenience/aesthetics thing. To avoid changing data_to_fit in rotate_and_project_data I would have to have the same if/else cases as done in this method in extract_data as well and I didn't want to duplicate code. 
That being said, I see your point. Should we discuss in person how to deal with this? For now I add a comment in extract_data to point out that data_to_fit will be modified later.

stephlazar replied:
I have now cleaned this up. Do you want to have another quick look or should we merge into Proj/S17?

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -739,19 +1136,22 @@ def cal_states_analysis(self):
                                                  default_value=True)
             if self.get_param_value('TwoD', default_value=False):
-                if self.rotation_type == 'global_PCA':
+                if self.rotation_type[qbn] == 'global_PCA':
```
use lower()? Further instance below.

stephlazar replied:
👍

stephlazar replied:
👍

chellings reviewed (commented):
Thank you, the new structure looks much cleaner. I already have some comments. I will have another look once the conflicts between Nathan's changes and this branch are resolved.

Commit:
Merge branch 'devel/S17/timedomain_classes' into enh/tda_cleanup

# Conflicts:
#	pycqed/analysis_v2/timedomain_analysis.py

Commit:
timedomain_analysis.py: Improvements and corrections in docstrings through MultiQubit_TimeDomain_Analysis

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -290,17 +400,271 @@ def extract_data(self):
+            - self.mospm (the meas_ibj_sweep_points_map)
+
+        Possible keyword arguments taken from metadata or options_dict:
+            - sp_filter: function to filter or process the sweep points
```
sp1d_filter?

maybe add a bit more explanation.

stephlazar replied:
👍

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -290,17 +400,271 @@ def extract_data(self):
+            - sweep_points_dict of the form {qbn: swpts_1d_array}
+            - swpts_1d_array from hard_sweep_params of the form
+            {sweep_param_name: {'values': swpts_1d_array, 'unit': unit_str}}
+            - self.raw_data_dict['hard_sweep_points'] created by the base class
```
General comment: there are cases where 1st dim sweep points are not hard sweep points (mixer calib, which will now also be analyzed with v2). At some point, we might want to clean up the naming conventions. (Not part of this PR.)

stephlazar replied:
👍

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -290,17 +400,271 @@ def extract_data(self):
+            {sweep_param_name: {'values': swpts_1d_array, 'unit': unit_str}}
+            - self.raw_data_dict['hard_sweep_points'] created by the base class
+                Only for this case, the self.data_filter is applied to the
+                sweep_points. TODO: understand why! (Steph, 12.08.2021)
```
open TODO item

stephlazar replied:
Now added a warning if the data_filter changes the len of the sweep points

chellings reviewed (commented):
A few more comments.

Commit:
Merge branch 'devel/S17/timedomain_classes' into enh/tda_cleanup

Commit:
MultiQubit_TimeDomain_Analysis: removes get_params_from_file

Commit:
MultiQubit_TimeDomain_Analysis: uses .lower() in string recognition of rotation_type

Commit:
Merge branch 'devel/S17/timedomain_classes' into enh/tda_cleanup

# Conflicts:
#	pycqed/analysis_v2/timedomain_analysis.py

Commit:
MultiQubit_TimeDomain_Analysis: bugfix: only add other cal states to projected_data_dict when not doing PCA

Commit:
MultiQubit_TimeDomain_Analysis: get_data_to_fit returns data_to_fit instead of creating self.data_to_fit

Commit:
MultiQubit_TimeDomain_Analysis: bugfix to get the transition name in get_transition_name from the unmodified data_to_fit when doing PCA

Commit:
MultiQubit_TimeDomain_Analysis: bugfix: make sure last_ge_pulses has the same size as self.qb_names if last_ge_pulses is not bool

Commit:
MultiQubit_TimeDomain_Analysis: remove keys from data_to_fit if they are not in self.qb_names

Commit:
MultiQubit_TimeDomain_Analysis: removes print statement

Commit:
MultiQubit_TimeDomain_Analysis: improves comment

Commit:
MultiQubit_TimeDomain_Analysis: improvements in comments

Commit:
MultiQubit_TimeDomain_Analysis: adds .lower() in rotation_type string recognition in rotate_and_project_data

Commit:
Merge branch 'devel/S17/timedomain_classes' into enh/tda_cleanup

Commit:
Merge branch 'devel/S17/timedomain_classes' into enh/tda_cleanup

Commit:
Merge branch 'devel/S17/timedomain_classes' into enh/tda_cleanup

Commit:
MultiQubit_TimeDomain_Analysis: fixes bug in get_cal_states_dict_for_rotation

Commit:
Merge branch 'devel/S17/timedomain_classes' into enh/tda_cleanup

Commit:
MultiQubit_TimeDomain_Analysis: cleanup of data_to_fit creation

- renames get_data_to_fit to _get_default_data_to_fit
- adds update_data_to_fit (updates based on rotation_type)
- removes storing_keys logic from rotate_and_project_data
- calls self._get_default_data_to_fit in self.get_transition_name

Commit:
MultiQubit_TimeDomain_Analysis: evaluate sp1d_filter in create_sweep_points_dict if specified as string + improve docstring

Commit:
Merge branch 'devel/S17/timedomain_classes' into enh/tda_cleanup

Commit:
MultiQubit_TimeDomain_Analysis: raise warning in fallback case in create_sweep_points_dict if the data_filter changes the length of the sweep points

Commit:
MultiQubit_TimeDomain_Analysis.get_cal_points: only raise warning/error if cal_points were not found

Commit:
MultiQubit_TimeDomain_Analysis.get_rotation_type: do not overwrite rotation_type if it was set to no_rotation

Commit:
QuDev_transmon: adds exp_metadata to calibrate_drive_mixer_carrier_model and calibrate_drive_mixer_skewness_model

Commit:
MixerCarrierAnalysis: fixes array shape bug that caused the analysis to crash for a random grid

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -694,49 +1007,146 @@ def unique(l):
 
-    def cal_states_analysis(self):
-        self.get_cal_data_points()
+        ! Note: self.data_to_fit is updated to reflect what was done here for
```
outdated comment?

stephlazar replied:
👍

Commit:
MultiQubit_TimeDomain_Analysis.prepare_projected_data_plot: adds comment

Commit:
MultiQubit_TimeDomain_Analysis.rotate_and_project_data: removed outdated comment

chellings reviewed (approved):
Looks good now. Thanks!
